lib/Firtool/Firtool.cpp

```diff
@@ -0,0 +1,246 @@
+        firrtl::createInferReadWritePass());
+
+  if (opt.replSeqMem)
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createLowerMemoryPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createPrefixModulesPass());
+
+  if (!opt.disableOptimization)
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createIMConstPropPass());
+
+  pm.addNestedPass<firrtl::CircuitOp>(firrtl::createAddSeqMemPortsPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createCreateSiFiveMetadataPass(
+      opt.replSeqMem, opt.replSeqMemCircuit, opt.replSeqMemFile));
+
+  pm.addNestedPass<firrtl::CircuitOp>(firrtl::createExtractInstancesPass());
+
+  // Run passes to resolve Grand Central features.  This should run before
+  // BlackBoxReader because Grand Central needs to inform BlackBoxReader where
+  // certain black boxes should be placed.  Note: all Grand Central Taps related
+  // collateral is resolved entirely by LowerAnnotations.
+  pm.addNestedPass<firrtl::CircuitOp>(
+      firrtl::createGrandCentralPass(opt.grandCentralInstantiateCompanionOnly));
+
+  // Read black box source files into the IR.
+  StringRef blackBoxRoot = opt.blackBoxRootPath.empty()
+                               ? llvm::sys::path::parent_path(inputFilename)
+                               : opt.blackBoxRootPath;
+  pm.nest<firrtl::CircuitOp>().addPass(
+      firrtl::createBlackBoxReaderPass(blackBoxRoot));
+
+  // Run SymbolDCE as late as possible, but before InnerSymbolDCE. This is for
+  // hierpathop's and just for general cleanup.
+  pm.addNestedPass<firrtl::CircuitOp>(mlir::createSymbolDCEPass());
+
+  // Run InnerSymbolDCE as late as possible, but before IMDCE.
+  pm.addPass(firrtl::createInnerSymbolDCEPass());
+
+  // The above passes, IMConstProp in particular, introduce additional
+  // canonicalization opportunities that we should pick up here before we
+  // proceed to output-specific pipelines.
+  if (!opt.disableOptimization) {
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        createSimpleCanonicalizerPass());
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        circt::firrtl::createRegisterOptimizerPass());
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createIMDeadCodeElimPass());
+  }
+
+  if (opt.emitOMIR)
+    pm.nest<firrtl::CircuitOp>().addPass(
+        firrtl::createEmitOMIRPass(opt.omirOutFile));
+
+  // Always run this, required for legalization.
+  pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+      firrtl::createMergeConnectionsPass(
+          !opt.disableAggressiveMergeConnections.getValue()));
+
+  if (!opt.disableOptimization)
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        firrtl::createVectorizationPass());
+
+  return success();
+}
+
+LogicalResult circt::populateLowFIRRTLToHW(mlir::PassManager &pm,
+                                           const FirtoolOptions &opt) {
+  if (opt.stripMuxPragmas == opt.addMuxPragmas) {
+    llvm::errs()
```

Comment:
Is there a diagnostic context we can assume here?  Would be nice to not have library code writing to the error stream.

Checking for validity on construction has benefits but separates that logic from the pipeline construction which I don't love.  Anyway this is fine as-is, just some thoughts.

Reply:
I originally added this condition and I should have had cleaned up more earilier:) I think it would be good time to remove this. Could you just drop `opt.stripMuxPragmas`?